### PR TITLE
v1.2.1 attendence

### LINKED ISSUE #5: Track attendance progress
As a user, I can track my student’s attendance, so I can take note of the tutorial participations of my students.

### LINKED ISSUE #10: Take attendance
As a user, I can take attendance through the same app. (so I can upload them in inetNUS).

### PR DESCRIPTION
**Pull request by YuunoKun**
Implements new Attendance feature together with Session/SessionList.
This PR does not guarantee the functionality of the Participate and Presence command, as it can only be completely done during integration (It needs to 'enter a session' which will have to acquire SessionName internally).

### CODE REVIEW

File: src/main/java/seedu/address/commons/core/Messages.java [hunk 1]
```diff
@@ -9,6 +9,7 @@
     public static final String MESSAGE_INVALID_COMMAND_FORMAT = "Invalid command format! \n%1$s";
     public static final String MESSAGE_INVALID_PERSON_DISPLAYED_INDEX = "The student index provided is invalid";
     public static final String MESSAGE_PERSONS_LISTED_OVERVIEW = "%1$d students listed!";
+    public static final String MESSAGE_SESSION_DOES_NOT_EXIST = "The session does not exist";
```
Comment (marcustw) [inline src/main/java/seedu/address/commons/core/Messages.java:12]: Perhaps missing a fullstop?

File: src/main/java/seedu/address/model/attendance/SessionDate.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    public static final String MESSAGE_CONSTRAINTS = "The date of a session should be in the form of dd/MM/yy "
+            + "corresponding to a valid date.\n"
+            + "Example : 10/2/2019";
+    public static final String VALIDATION_REGEX = "[0-9]{1,2}/[0-9]{1,2}/[0-9]{4,4}";
+    public static final String PLACEHOLDER_DATE = "1/1/2020";
+
+    public final LocalDate value;
+
+    /**
+     * Constructs an {@code SessionDate}.
```
Comment (marcustw) [inline src/main/java/seedu/address/model/attendance/SessionDate.java]: Perhaps "Constructs a {@code SessionDate}" ?

File: src/main/java/seedu/address/model/attendance/SessionDate.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+    /**
+     * Returns if a given string is a valid date.
+     */
+    public static boolean isValidSessionDate(String date) {
+        return date.matches(VALIDATION_REGEX);
+    }
+
+    /**
+     * Standardize date string into the format of dd/MM/yyyy.
```
Comment (marcustw) [inline src/main/java/seedu/address/model/attendance/SessionDate.java:43]: Perhaps "Standardizes" instead of "Standardize"?

File: src/main/java/seedu/address/model/attendance/exceptions/SessionNotFoundException.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package seedu.address.model.attendance.exceptions;
+
+/**
+ * Signals that the operation will result in duplicate Classes (Classes are considered duplicates if they have the same
+ * date).
+ */
+public class SessionNotFoundException extends RuntimeException{}
```
Comment (marcustw) [inline src/main/java/seedu/address/model/attendance/exceptions/SessionNotFoundException.java:7]: Is there a missing space between RunTimeException and "{}"?

File: src/main/java/seedu/address/storage/JsonAdaptedAttributes.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+import seedu.address.commons.exceptions.IllegalValueException;
+import seedu.address.model.attendance.Attributes;
+import seedu.address.model.attendance.Participation;
+import seedu.address.model.attendance.Presence;
+
+/**
+ * Jackson-friendly version of {@link Attributes}.
+ */
+public class JsonAdaptedAttributes {
```
Comment (marcustw) [inline src/main/java/seedu/address/storage/JsonAdaptedAttributes.java:16]: Awesome!

**Review by marcustw — COMMENTED** [review]
Excellent work Sheng Yang!

File: src/main/java/seedu/address/logic/commands/ClearSessionCommand.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+/**
+ * Clears the session list.
+ */
+public class ClearSessionCommand extends Command {
+
+    public static final String COMMAND_WORD = "clearsession";
+    public static final String MESSAGE_SUCCESS = "The session in the current session list has been cleared";
+
+
```
Comment (erisjacey) [inline src/main/java/seedu/address/logic/commands/ClearSessionCommand.java:15]: Should this unncessary extra line break be removed?

File: src/main/java/seedu/address/logic/parser/CliSyntax.java [hunk 7]
```diff
@@ -10,5 +10,7 @@
     public static final Prefix PREFIX_MATRICULATION = new Prefix("m/");
     public static final Prefix PREFIX_EMAIL = new Prefix("e/");
     public static final Prefix PREFIX_TAG = new Prefix("t/");
+    public static final Prefix PREFIX_SESSIONNAME = new Prefix("s/");
+    public static final Prefix PREFIX_SESSIONDATE = new Prefix("d/");
```
Comment (erisjacey) [inline src/main/java/seedu/address/logic/parser/CliSyntax.java:14]: Maybe these could be renamed to `PREFIX_SESSION_NAME` and `PREFIX_SESSION_DATE`?

File: src/main/java/seedu/address/model/attendance/SessionDate.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+import java.time.format.DateTimeFormatter;
+
+
+/**
+ * Represents a date of a session in the session list.
+ * Guarantees: immutable; is valid as declared in {@link #isValidSessionDate(String)}
+ */
+public class SessionDate implements Comparable<SessionDate> {
+
+    public static final String MESSAGE_CONSTRAINTS = "The date of a session should be in the form of dd/MM/yy "
```
Comment (erisjacey) [inline src/main/java/seedu/address/model/attendance/SessionDate.java]: Shouldn't the form be `dd/MM/yyyy`?

File: src/main/java/seedu/address/model/attendance/SessionList.java [hunk 9]
```diff
@@ -47,19 +103,110 @@ public boolean contains(Session toCheck) {
...
+            for (Session s: ((SessionList) other).getSessions()) {
+                testArray2[j] = s;
+                j++;
+            }
+
+            for (int k = 0; k < sessions.size(); k++) {
+                if (!testArray1[k].isSameSession(testArray2[k])) {
+                    return false;
+                }
+            }
```
Comment (erisjacey) [inline src/main/java/seedu/address/model/attendance/SessionList.java:200]: Maybe some comments could be added here to more clearly explain the thought process?

File: src/main/java/seedu/address/storage/JsonAdaptedSession.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+import seedu.address.model.attendance.Session;
+import seedu.address.model.attendance.SessionDate;
+import seedu.address.model.attendance.SessionName;
+
+/**
+ * Jackson-friendly version of {@link Session}.
+ */
+public class JsonAdaptedSession {
+
+
```
Comment (erisjacey) [inline src/main/java/seedu/address/storage/JsonAdaptedSession.java:23]: Unnecessary additional line break?

File: src/main/java/seedu/address/storage/JsonAdaptedSession.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+        if (!SessionName.isValidSessionName(sessionName)) {
+            throw new IllegalValueException(SessionName.MESSAGE_CONSTRAINTS);
+        }
+        final SessionName name = new SessionName(sessionName);
+
+        if (sessionDate == null) {
+            throw new IllegalValueException(SessionDate.MESSAGE_CONSTRAINTS);
+        }
+        if (!SessionDate.isValidSessionDate(sessionDate)) {
+            throw new IllegalValueException(SessionDate.MESSAGE_CONSTRAINTS);
```
Comment (erisjacey) [inline src/main/java/seedu/address/storage/JsonAdaptedSession.java:88]: Maybe these two `if` blocks could be merged into one?

File: src/main/java/seedu/address/storage/JsonAdaptedSession.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+
+            studentList.put(Integer.parseInt(attr.getAttributeIndex()), attributes);
+        }
+
+        if (sessionName == null) {
+            throw new IllegalValueException(SessionName.MESSAGE_CONSTRAINTS);
+        }
+        if (!SessionName.isValidSessionName(sessionName)) {
+            throw new IllegalValueException(SessionName.MESSAGE_CONSTRAINTS);
+        }
```
Comment (erisjacey) [inline src/main/java/seedu/address/storage/JsonAdaptedSession.java:81]: Maybe these two `if` blocks could be merged into one?

File: src/test/java/seedu/address/logic/commands/CommandTestUtil.java [hunk 13]
```diff
@@ -55,13 +61,51 @@
...
+    // ---------------------- Session Testing -------------------------------
+    public static final String VALID_SESSIONNAME_REC = "recitation1";
+    public static final String VALID_SESSIONNAME_CON = "midterm consultation";
+    public static final String VALID_SESSIONDATE_REC = "12/7/2017";
+    public static final String VALID_SESSIONDATE_CON = "27/11/2015";
+
+    public static final String VALID_SESSIONNAME_REC_DESC = " " + PREFIX_SESSIONNAME + VALID_SESSIONNAME_REC;
+    public static final String VALID_SESSIONNAME_CON_DESC = " " + PREFIX_SESSIONNAME + VALID_SESSIONNAME_CON;
+    public static final String VALID_SESSIONDATE_REC_DESC = " " + PREFIX_SESSIONDATE + VALID_SESSIONDATE_REC;
+    public static final String VALID_SESSIONDATE_CON_DESC = " " + PREFIX_SESSIONDATE + VALID_SESSIONDATE_CON;
```
Comment (erisjacey) [inline src/test/java/seedu/address/logic/commands/CommandTestUtil.java:74]: Would it make more sense if all instances of `SESSIONNAME` got changed to `SESSION_NAME`? (same for `SESSIONDATE` and `SESSION_DATE`?

File: src/test/java/seedu/address/logic/commands/EditSessionCommandTest.java [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+import seedu.address.model.Model;
+import seedu.address.model.ModelManager;
+import seedu.address.model.UserPrefs;
+import seedu.address.model.attendance.Session;
+import seedu.address.model.attendance.SessionName;
+import seedu.address.testutil.EditSessionDescriptorBuilder;
+import seedu.address.testutil.ModelManagerBuilder;
+
+
+
```
Comment (erisjacey) [inline src/test/java/seedu/address/logic/commands/EditSessionCommandTest.java:36]: Unnecssary whitespaces?

File: src/test/java/seedu/address/logic/parser/AddSessionCommandParserTest.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+import org.junit.jupiter.api.Test;
+
+import seedu.address.logic.commands.AddSessionCommand;
+import seedu.address.model.attendance.Session;
+import seedu.address.model.attendance.SessionDate;
+import seedu.address.model.attendance.SessionName;
+import seedu.address.testutil.SessionBuilder;
+
+
+
```
Comment (erisjacey) [inline src/test/java/seedu/address/logic/parser/AddSessionCommandParserTest.java:26]: Unnecssary whitespaces?

File: src/test/java/seedu/address/logic/parser/DeleteSessionCommandParserTest.java [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+import static seedu.address.logic.parser.CommandParserTestUtil.assertParseSuccess;
+import static seedu.address.testutil.TypicalSessions.SESSIONNAME_WEEK_ONE;
+import static seedu.address.testutil.TypicalSessions.SESSIONNAME_WEEK_ONE_STRING;
+
+import org.junit.jupiter.api.Test;
+
+import seedu.address.logic.commands.DeleteSessionCommand;
+
+
+
```
Comment (erisjacey) [inline src/test/java/seedu/address/logic/parser/DeleteSessionCommandParserTest.java:12]: Unnecessary whitespaces?

File: src/test/java/seedu/address/logic/parser/EditSessionCommandParserTest.java [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+
+import org.junit.jupiter.api.Test;
+
+import seedu.address.logic.commands.EditSessionCommand;
+import seedu.address.model.attendance.SessionDate;
+import seedu.address.model.attendance.SessionName;
+import seedu.address.testutil.EditSessionDescriptorBuilder;
+
+
+
```
Comment (erisjacey) [inline src/test/java/seedu/address/logic/parser/EditSessionCommandParserTest.java:25]: Unnecessary whitespaces?

File: src/test/java/seedu/address/logic/parser/ParticipateCommandParserTest.java [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+import static seedu.address.logic.parser.CommandParserTestUtil.assertParseSuccess;
+
+import org.junit.jupiter.api.Test;
+
+import seedu.address.logic.commands.ParticipateCommand;
+import seedu.address.model.attendance.IndexRange;
+import seedu.address.model.attendance.SessionName;
+
+
+
```
Comment (erisjacey) [inline src/test/java/seedu/address/logic/parser/ParticipateCommandParserTest.java:15]: Unnecessary whitespaces?

File: src/test/java/seedu/address/logic/parser/PresenceCommandParserTest.java [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+import static seedu.address.logic.parser.CommandParserTestUtil.assertParseFailure;
+import static seedu.address.logic.parser.CommandParserTestUtil.assertParseSuccess;
+
+import org.junit.jupiter.api.Test;
+
+import seedu.address.logic.commands.PresenceCommand;
+import seedu.address.model.attendance.IndexRange;
+import seedu.address.model.attendance.SessionName;
+
+
```
Comment (erisjacey) [inline src/test/java/seedu/address/logic/parser/PresenceCommandParserTest.java:14]: Unnecessary extra line break?

File: src/test/java/seedu/address/model/ModelManagerTest.java [hunk 20]
```diff
@@ -18,6 +19,7 @@
 import seedu.address.model.person.NameContainsKeywordsPredicate;
 import seedu.address.testutil.AddressBookBuilder;
 
+
```
Comment (erisjacey) [inline src/test/java/seedu/address/model/ModelManagerTest.java:22]: Unnecessary line break?

File: src/test/java/seedu/address/model/attendance/IndexRangeTest.java [hunk 21]
```diff
@@ -0,0 +1,42 @@
+package seedu.address.model.attendance;
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static seedu.address.testutil.Assert.assertThrows;
+
+import org.junit.jupiter.api.Test;
+
+
```
Comment (erisjacey) [inline src/test/java/seedu/address/model/attendance/IndexRangeTest.java:9]: Unnecessary line break?

File: src/test/java/seedu/address/model/attendance/SessionDateTest.java [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+import static seedu.address.testutil.TypicalSessions.SESSIONDATE_WEEK_ONE;
+import static seedu.address.testutil.TypicalSessions.SESSIONDATE_WEEK_ONE_STRING;
+import static seedu.address.testutil.TypicalSessions.SESSIONDATE_WEEK_THREE_STRING;
+import static seedu.address.testutil.TypicalSessions.SESSIONDATE_WEEK_TWO;
+
+import java.time.format.DateTimeParseException;
+
+import org.junit.jupiter.api.Test;
+
+
```
Comment (erisjacey) [inline src/test/java/seedu/address/model/attendance/SessionDateTest.java:15]: Unnecessary line break?

File: src/test/java/seedu/address/storage/JsonAdaptedSessionTest.java [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.List;
+
+import org.junit.jupiter.api.Test;
+
+import seedu.address.commons.exceptions.IllegalValueException;
+import seedu.address.model.attendance.SessionDate;
+import seedu.address.model.attendance.SessionName;
+
+
+
```
Comment (erisjacey) [inline src/test/java/seedu/address/storage/JsonAdaptedSessionTest.java:21]: Unnecessary whitespaces?

File: src/test/java/seedu/address/storage/JsonSerializableSessionListTest.java [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+
+import seedu.address.commons.exceptions.IllegalValueException;
+import seedu.address.commons.util.JsonUtil;
+import seedu.address.model.attendance.SessionList;
+import seedu.address.testutil.TypicalSessions;
+
+
+
+
+
```
Comment (erisjacey) [inline src/test/java/seedu/address/storage/JsonSerializableSessionListTest.java:19]: Unnecessary whitespaces?

**Review by erisjacey — COMMENTED** [review]
Amazing job! Just some minor nitpicks with regard to code styling (mostly just unnecessary line breaks/white spaces, a few naming alternatives to consider).

Stellar job nonetheless 😄 This is really top-notch effort.

File: src/main/java/seedu/address/commons/core/Messages.java (hunk 1 above)
Comment (YuunoKun) [reply src/main/java/seedu/address/commons/core/Messages.java:12]: Not sure about this, the punctuation isn't consistent for other message strings too. I guess we can clean this up after integration.

File: src/main/java/seedu/address/model/ModelManager.java [hunk 25]
```diff
@@ -95,14 +103,23 @@ public boolean hasPerson(Person person) {
...
-    public void deletePerson(Person target) {
+    public void deletePerson(Person target, Index id) {
         addressBook.removePerson(target);
+        sessionList.updatePersonList(addressBook.getPersonList());
+        sessionList.updateAllSessionsAfterDelete(id);
     }
 
     @Override
     public void addPerson(Person person) {
+        for (Session s: sessionList.getSessions()) {
```
Comment (CodingCookieRookie) [inline src/main/java/seedu/address/model/ModelManager.java]: Should we remove this and use run in debug mode when its finalised

File: src/test/java/seedu/address/logic/commands/AddSessionCommandTest.java [hunk 26]
```diff
@@ -0,0 +1,250 @@
...
+public class AddSessionCommandTest {
+
+    @Test
+    public void constructor_nullSession_throwsNullPointerException() {
+        assertThrows(NullPointerException.class, () -> new AddSessionCommand(null));
+    }
+
+    @Test
+    public void execute_sessionAcceptedByModel_addSuccessful() throws Exception {
+        ModelStubAcceptingSessionAdded modelStub = new ModelStubAcceptingSessionAdded();
```
Comment (CodingCookieRookie) [inline src/test/java/seedu/address/logic/commands/AddSessionCommandTest.java:40]: I like your usage of stubs!

File: src/main/java/seedu/address/model/attendance/IndexRange.java [hunk 27]
```diff
@@ -0,0 +1,104 @@
...
+        checkArgument(isValidIndexRange(range), MESSAGE_CONSTRAINTS);
+        value = range;
+        fillParams(range);
+    }
+
+    /**
+     * Returns if a given string is a valid index range.
+     */
+    public static boolean isValidIndexRange(String test) {
+        return test.matches(VALIDATION_REGEX_1) || test.matches(VALIDATION_REGEX_2);
```
Comment (CodingCookieRookie) [inline src/main/java/seedu/address/model/attendance/IndexRange.java:42]: Pattern.matches throw some exceptions when dealing with some corner cases like. Perhaps String.matches() do as well?

File: src/main/java/seedu/address/model/attendance/Session.java [hunk 28]
```diff
@@ -17,16 +16,26 @@
  */
 public class Session implements Comparable<Session> {
 
-    private final LocalDate date;
+    private final SessionName sessionName;
+    private final SessionDate sessionDate;
     private final Map<Integer, Attributes> studentList;
```
Comment (CodingCookieRookie) [inline src/main/java/seedu/address/model/attendance/Session.java:21]: Perhaps studentAttributesList would be better here?

File: src/main/java/seedu/address/commons/core/Messages.java (hunk 1 above)
Comment (marcustw) [reply src/main/java/seedu/address/commons/core/Messages.java:12]: Sure thing!

**Review by CodingCookieRookie — APPROVED** [review]
LGTM
